language/evm/move-ethereum-annotator/src/util.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::LRU_CACHE_SIZE;
+use anyhow::{anyhow, bail, Result};
+use ethabi::{Event, Function, Token};
+use ethereum_types::U256;
+use lru::LruCache;
+use move_binary_format::CompiledModule;
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::Identifier,
+    language_storage::{ModuleId, StructTag},
+    resolver::ModuleResolver,
+    value::MoveValue,
+};
+use move_ethereum_abi::abi_move_type::{ABIMoveSignature, ABI_ETHER_MOVE_KEY};
+use std::{collections::HashMap, convert::TryFrom, str::FromStr};
+
+pub(crate) struct ModuleCache<R: ModuleResolver> {
+    resolver: R,
+    cache: LruCache<ModuleId, CacheEntry>,
+    // TODO: Review the selector map's value type. Will function name be sufficient here or could there be hash collision issues?
+    selector_cache: LruCache<[u8; 4], String>,
```

Comment:
I think we should consider the hash collision issues. If we define selector_cache as LruCache<[u8; 4], Vec<String>> can resolve this?

Reply:
How a `Vec` here could help the issue? The attack vector I have on top of my mind is following: 
module A has an api that has signature `Foo(arg1, arg2)` and module B has an api that has signature `Bar(arg1, arg2)` where the selector of `Foo` and `Bar` collides (it's easy to come up with sth like that as we only take the top 4 bytes for the selector). Then during the transaction, when we invoke `A::Foo` it could resolve to `A::Bar` because of the collision. Maybe a selector cache per module will resolve this issue (but could explode the cache size potentially).

Reply:
if use Vec<String> to keep the function name, if the resolved result `A::Bar` does not exist in the A module, we should try to fetch the module again, and mapping short_signature with `["Foo", "Bar"].

Maybe a selector cache per module is more suitable.